File: WoofIOS/SwiftPackages/Presentation/Common/Sources/PresentationCommon/ViewModel/Base/BaseLifecycleViewModel.swift
import Combine

open class BaseLifecycleViewModel: ObservableObject {
    public var cancellables = Set<AnyCancellable>()
    private var didViewAppear = false

    public init() {}

    public final func onViewAppear() {
        if !didViewAppear {
            onFirstAppear()
            didViewAppear = true
        }
        onEveryAppear()
    }

    open func onFirstAppear() {}

    open func onEveryAppear() {}
}

File: WoofAndroid/presentation/common/src/main/kotlin/com/perrystreet/woof/presentation/common/viewmodel/base/BaseLifecycleViewModel.kt
package com.perrystreet.woof.presentation.common.viewmodel.base

import androidx.lifecycle.ViewModel
import io.reactivex.rxjava3.disposables.CompositeDisposable

abstract class BaseLifecycleViewModel : ViewModel() {
    protected val disposables = CompositeDisposable()
    private var didViewAppear = false

    override fun onCleared() {
        super.onCleared()
        disposables.dispose()
    }

    fun onViewAppear() {
        if (!didViewAppear) {
            onFirstAppear()
            didViewAppear = true
        }
        onEveryAppear()
    }

    protected open fun onFirstAppear() = Unit

    protected open fun onEveryAppear() = Unit
}
